src/Dapplo.Microsoft.Extensions.Hosting.Wpf/HostBuilderWpfExtensions.cs

```diff
@@ -73,35 +77,51 @@ public static IHostBuilder ConfigureWpf(this IHostBuilder hostBuilder, Action<IW
                     serviceCollection.AddSingleton(serviceProvider => new WpfThread(serviceProvider));
                     serviceCollection.AddHostedService<WpfHostedService>();
                 }
-                configureAction?.Invoke(wpfContext);
+                wpfBuilder.ConfigureContextAction?.Invoke(wpfContext);
             });
-            return hostBuilder;
-        }
 
-        /// <summary>
-        /// Configure an WPF application
-        /// </summary>
-        /// <param name="hostBuilder">IHostBuilder</param>
-        /// <param name="configureAction">Action to configure the Application</param>
-        /// <typeparam name="TView">
-        ///     Type for the view, must derive from Window.
-            /// If derived from IWpfShell, it automatically shown at start</typeparam>
-        /// <returns>IHostBuilder</returns>
-        public static IHostBuilder ConfigureWpf<TView>(this IHostBuilder hostBuilder, Action<IWpfContext> configureAction = null) where TView : Window
-        {
-            hostBuilder.ConfigureWpf(configureAction);
-            hostBuilder.ConfigureServices((hostBuilderContext, serviceCollection) =>
+            if (wpfBuilder.ApplicationType != null)
             {
-                serviceCollection.AddSingleton<TView>();
+                // Check if the registered application does inherit System.Windows.Application
+                var baseApplicationType = typeof(Application);
+                if (!baseApplicationType.IsAssignableFrom(wpfBuilder.ApplicationType))
+                {
+                    throw new ArgumentException("The registered Application type inherit System.Windows.Application", nameof(configureDelegate));
+                }
 
-                // Check if it also implements IWpfShell so we can register it as this
-                var viewType = typeof(TView);
-                var shellInterfaceType = typeof(IWpfShell);
-                if (shellInterfaceType.IsAssignableFrom(viewType))
+                hostBuilder.ConfigureServices((hostBuilderContext, serviceCollection) =>
                 {
-                    serviceCollection.AddSingleton(shellInterfaceType, serviceProvider => serviceProvider.GetRequiredService<TView>());
+                    serviceCollection.AddSingleton(wpfBuilder.ApplicationType);
+
+                    if (wpfBuilder.ApplicationType != baseApplicationType)
+                    {
+                        serviceCollection.AddSingleton(serviceProvider => (Application)serviceProvider.GetRequiredService(wpfBuilder.ApplicationType));
+                    }
+                });
+            }
+
+            if (wpfBuilder.MainWindowType != null)
```

Comment:
Make it possible to supply multiple, from my point of view a shell doesn't need to be limited to one window.